llama-index-integrations/instrumentation/llama-index-instrumentation-agentops/llama_index/instrumentation/agentops/base.py

```diff
@@ -0,0 +1,238 @@
+        return super().new_span(id_, bound_args, instance, parent_span_id, **kwargs)
+
+    def prepare_to_exit_span(
+        self,
+        id_: str,
+        bound_args: BoundArguments,
+        instance: Optional[Any] = None,
+        result: Optional[Any] = None,
+        **kwargs: Any,
+    ) -> SimpleSpan:
+        self._shared_handler_state.remove_span_id(id_)
+        return super().prepare_to_exit_span(id_, bound_args, instance, result, **kwargs)
+
+    def prepare_to_drop_span(
+        self,
+        id_: str,
+        bound_args: BoundArguments,
+        instance: Optional[Any] = None,
+        err: Optional[BaseException] = None,
+        **kwargs: Any,
+    ) -> SimpleSpan:
+        if err:
+            # Associate this exception to the parent span, so that it will know to ignore it
+            parent_span_id = self._shared_handler_state.span_parent.get(id_, None)
+            if parent_span_id:
+                if parent_span_id in self._shared_handler_state.span_exception:
+                    self._shared_handler_state.span_exception[parent_span_id].add(err)
+                else:
+                    self._shared_handler_state.span_exception[parent_span_id] = {err}
+
+            # If this exception hasn't yet been associated to this span, record it
+            if (
+                id_ not in self._shared_handler_state.span_exception
+                or err not in self._shared_handler_state.span_exception[id_]
+            ):
+                self._ao_client.record(ErrorEvent(details=str(err)))
+
+        self._shared_handler_state.remove_span_id(id_)
+        return super().prepare_to_drop_span(id_, bound_args, instance, err, **kwargs)
+
+
+class AgentOpsEventHandler(BaseEventHandler):
+    _shared_handler_state: AgentOpsHandlerState = PrivateAttr()
+    _ao_client: AOClient = PrivateAttr()
+
+    def __init__(
+        self, shared_handler_state: AgentOpsHandlerState, ao_client: AOClient
+    ) -> None:
+        self._shared_handler_state = shared_handler_state
+        self._ao_client = ao_client
+        super().__init__()
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "AgentOpsEventHandler"
+
+    def handle(self, event: BaseEvent) -> None:
+        # We only track chat events that are emitted while using an agent
+        is_agent_chat_event = self._shared_handler_state.check_is_agent_chat_span(
+            event.span_id
+        )
+
+        if isinstance(event, AgentRunStepStartEvent):
+            self._shared_handler_state.is_agent_chat_span[event.span_id] = True
+
+        if isinstance(event, LLMChatStartEvent) and is_agent_chat_event:
+            self._shared_handler_state.agent_chat_start_event[event.span_id] = event
+            model = event.model_dict["model"] if "model" in event.model_dict else None
+            prompt = messages_to_prompt(event.messages)
+            self._ao_client.record(LLMEvent(model=model, prompt=prompt))
```

Comment:
Can we also include `prompt_tokens` and `completion_tokens` if available? 

AgentOps does server-side token calculation as a fallback, but it's always better to have the tokens directly from the response when they're provided

Reply:
They aren't easily available, especially before an LLM call is made. Could try adding after the call is made

Reply:
Done